k-distribution/src/main/scripts/lib/pyk/pyk/kastManip.py

```diff
@@ -212,28 +266,59 @@ def addOccurance(k):
     return occurances
 
 
-def countVarOccurances(kast, numOccurances=None):
-    """Count the number of occurances of each variable in a proof.
+def extract_lhs(term: KInner) -> KInner:
+    return traverseTopDown(term, if_ktype(KRewrite, lambda rw: rw.lhs))
 
-    -   Input: Kast term.
-    -   Output: Map of variable names to their number of occurances.
-    """
-    numOccurances = {} if numOccurances is None else numOccurances
 
-    def _getNumOccurances(_kast):
-        if type(_kast) is KVariable:
-            vName = _kast.name
-            if vName in numOccurances:
-                numOccurances[vName] += 1
-            else:
-                numOccurances[vName] = 1
+def extract_rhs(term: KInner) -> KInner:
+    return traverseTopDown(term, if_ktype(KRewrite, lambda rw: rw.rhs))
+
+
+def count_vars(term: KInner) -> Counter:
+    counter: Counter = Counter()
+
+    def count(term: KInner) -> None:
+        if type(term) is KVariable:
+            counter[term.name] += 1
 
-    collectBottomUp(kast, _getNumOccurances)
-    return numOccurances
+    collectBottomUp(term, count)
+    return counter
+
+
+def count_rhs_vars(term: KInner) -> Counter:
+    def recur(term: KInner, *, rhs=False) -> Counter:
+        if type(term) is KVariable:
+            return Counter(term.name) if rhs else Counter()
+        if type(term) is KRewrite:
+            return recur(term.rhs, rhs=True)
+        if type(term) is KApply:
+            return sum((recur(t, rhs=rhs) for t in term.args), Counter())
+        if type(term) is KSequence:
+            return sum((recur(t, rhs=rhs) for t in term.items), Counter())
+        return Counter()
+    return recur(term)
```

Comment:
I added this function so that variables can be prefixed with `_` and `?` accordingly based on the number of their occurrences in either side in claims.

The problem I see with a solution based on `extract_lhs` and `extract_rhs` is that it cannot differentiate between certain terms that have to handled differently. Consider e.g. `<cell> V </cell>` and `<cell> V => 2 *Int V </cell>`, where the first should be rewritten as `<cell> _V </cell>` while the second is fine as is.